src/coreclr/vm/customattribute.cpp

```diff
@@ -1180,3 +835,316 @@ ARG_SLOT COMCustomAttribute::GetDataFromBlob(Assembly *pCtorAssembly,
+                            break;
+                        BOOL bObjectCreated = FALSE;
+                        TypeHandle th = pSig->GetLastTypeHandleThrowing();
+                        if (th.IsArray())
+                            // get the array element
+                            th = th.GetArrayElementTypeHandle();
+                        ARG_SLOT data = GetDataFromBlob(pCtorMD->GetAssembly(), (CorSerializationType)type, th, &pBlob, pEndBlob, pModule, &bObjectCreated);
+                        if (bObjectCreated)
+                            argToProtect[i] = ArgSlotToObj(data);
+                        else
+                            args[i] = data;
+                    }
+                }
+                GCPROTECT_END();
+
+                // We have borrowed the signature from MethodDescCallSite. We have to put it back into the initial position
+                // because of that's where MethodDescCallSite expects to find it below.
+                pSig->Reset();
+
+                for (i = 1; i < cArgs; i++)
+                {
+                    if (argToProtect[i] != NULL)
+                    {
+                        _ASSERTE(args[i] == NULL);
+                        args[i] = ObjToArgSlot(argToProtect[i]);
+                    }
+                }
+            }
+        }
+        args[0] = ObjToArgSlot(argToProtect[0]);
+
+        if (i != cArgs)
+            COMPlusThrow(kCustomAttributeFormatException);
+
+        // check if there are any named properties to invoke,
+        // if so set the by ref int passed in to point
+        // to the blob position where name properties start
+        *pcNamedArgs = 0;
+
+        if (pBlob && pBlob != pEndBlob)
+        {
+            if (pBlob + 2 > pEndBlob)
+                COMPlusThrow(kCustomAttributeFormatException);
+
+            *pcNamedArgs = GET_UNALIGNED_VAL16(pBlob);
+
+            pBlob += 2;
+        }
+
+        *ppBlob = pBlob;
+
+        if (*pcNamedArgs == 0 && pBlob != pEndBlob)
+            COMPlusThrow(kCustomAttributeFormatException);
+
+        // make the invocation to the ctor
+        gc.ca = ArgSlotToObj(args[0]);
+        if (pCtorMD->GetMethodTable()->IsValueType())
+            args[0] = PtrToArgSlot(OBJECTREFToObject(gc.ca)->UnBox());
+
+        ctorCallSite.CallWithValueTypes(args);
+
+        result.Set(gc.ca);
+    }
+    GCPROTECT_END();
+    END_QCALL;
+}
+
+extern "C" void QCALLTYPE CustomAttribute_CreatePropertyOrFieldData(
+    QCall::ModuleHandle pModule,
+    BYTE** ppBlobStart,
+    BYTE* pBlobEnd,
+    QCall::StringHandleOnStack pName,
+    bool* pbIsProperty,
+    QCall::ObjectHandleOnStack pType,
+    QCall::ObjectHandleOnStack pValue)
+{
+    QCALL_CONTRACT;
+
+    BEGIN_QCALL;
+
+    BYTE* pBlob = *ppBlobStart;
+
+    GCX_COOP();
+
+    struct
+    {
+        STRINGREF Name;
+        OBJECTREF Type;
+        OBJECTREF Value;
+    } gc;
+    gc.Name = NULL;
+    gc.Type = NULL;
+    gc.Value = NULL;
+    GCPROTECT_BEGIN(gc);
+
```

Comment:
Nit: This GCPROTECT is unnecessary. You can set the ObjectHandleOnStack instead of catching the values in GCPROTECTed locals.

Reply:
Agree. I prefer this model because some members are checked in asserts and seeing a `.Get()` looks odd. Additionally having everything hanging off an explicit `gc` type makes lifetime explicit. I will take another look though.

Reply:
I've removed the `Name` and `Type`. The `Value` is a bit more difficult as we are passing the protected address to another function. Getting the indirection out of an `QCall::ObjectHandleOnStack` is going to be annoying. Let me know if you have any concerns.

Reply:
> Getting the indirection out of an QCall::ObjectHandleOnStack is going to be annoying

I agree. It does actually work with the OBJECTREF assert infrastructure: https://github.com/dotnet/runtime/blob/bf828d87a7cfdff5c4c7f219ef77b6b4c0be04cf/src/coreclr/vm/qcall.h#L227-L229

If we wanted, the way to get rid of the last one would be to change the function to return the OBJECTREF and GCPROJECT inside that function instead.

Reply:
That is a good idea. Let me change to that approach.

Reply:
Done.